src/ex2/main.js

```diff
@@ -0,0 +1,148 @@
+        onCheckUncheckTask: (taskTxt) => this.onCheckUncheckTask(taskTxt),
+        onFinishedAll: () => this.onFinishedAll(),
+        onClearAll: () => this.onClearAll(),
+      }
+    );
+  }
+
+  // Add task will be clickable only for a non empty/"only spaces" task
+  onTaskInput() {
+    if (!/^\s*$/.test(this.addTaskInput.value)) {
+      this.addTaskBtn.classList.add(validTask);
+    } else if (/^\s*$/.test(this.addTaskInput.value)) {
+      this.addTaskBtn.classList.remove(validTask);
+    }
+  }
+
+  clearAddTaskInput() {
+    this.addTaskInput.value = "";
+    this.addTaskBtn.classList.remove(validTask);
+  }
+
+  onAddTask(taskId) {
+    // Render tasks
+    this.todoList.todos.forEach((todo) => {
+      if (!this.diplayedTasksSet.has(todo.id)) {
+        document
+          .querySelector(newTasks)
+          .appendChild(this.createTask(todo));
+        this.diplayedTasksSet.add(todo.id);
+      }
+    });
+
+    this.clearAddTaskInput();
+    this.updateTasksLeft();
+
+    // list is empty and first task was added -> hide 'Done and Done' and show footer
+    if (this.todoList.todos.length === 1) {
+      this.toggleFinishedAllAndFooter();
+    }
+
+    const input = document.querySelector(`#task-${taskId}`);
+    input.addEventListener("click", (event) => this.onTaskPress(event, taskId));
+  }
+
+  onCheckUncheckTask(taskTxtElem) {
+    this.updateTasksLeft();
+    // Crossing/uncrossing a done task's text
+    taskTxtElem.classList.toggle("done-task-txt");
+  }
+
+  onTaskPress(event, taskId) {
+    const isTaskExist = document.querySelector(`#task-${taskId}`);
+    if (isTaskExist) {
+      const wasClickedOnTaskValue =
+        !document
+          .querySelector(`#task-checkbox-${taskId}`)
+          .contains(event.target) &&
+        !document
+          .querySelector(`#trash-button-${taskId}`)
+          .contains(event.target);
+      if (wasClickedOnTaskValue) {
+        alert(document.querySelector(`#task-txt-${taskId}`).innerText);
+      }
+    }
+  }
+
+  onTaskDelete(todoToDelete) {
+    document.querySelector(`#task-${todoToDelete.id}`).remove();
+    this.diplayedTasksSet.delete(todoToDelete.id);
+    this.updateTasksLeft();
+  }
+
+  onClearAll() {
+    const nodes = document.querySelector(newTasks).childNodes;
+    for (let i = nodes.length - 1; i >= 0; i--) {
+      nodes[i].remove();
+    }
+    this.toggleFinishedAllAndFooter();
+  }
+
+  // Tasks left
+  updateTasksLeft() {
+    const tasksLeft = this.todoList.todos.reduce(
+      (currTasksLeft, todo) =>
+        todo.toBeDone ? currTasksLeft + 1 : currTasksLeft,
+      0
+    );
+    const tasksLeftElem = document.querySelector(".tasks-left");
+    tasksLeftElem.innerText = `Keep Grinding! You got ${tasksLeft} to go`;
+  }
+
+  createTask({ id: taskId, value }) {
+    const task = document.createElement("div");
+    task.classList.add("task-container");
+    task.id = `task-${taskId}`;
+    task.innerHTML = `
```

Comment:
there's a better way to do it. it's not a good convention to add html code inside js

Reply:
By using createElement?